mountpoint-s3/src/data_cache/multilevel_cache.rs

```diff
@@ -0,0 +1,323 @@
+            bucket: bucket.to_string(),
+            part_size: PART_SIZE,
+            enable_backpressure: true,
+            initial_read_window_size: PART_SIZE,
+            ..Default::default()
+        };
+        let client = MockClient::new(config);
+        (
+            client.clone(),
+            ExpressDataCache::new(
+                bucket,
+                client,
+                "unique source description",
+                BLOCK_SIZE,
+                EXPRESS_CACHE_MAX_OBJECT_SIZE,
+            ),
+        )
+    }
+
+    #[tokio::test]
+    async fn test_put_in_both_caches() {
+        let disk_cache = default_disk_cache();
+        let (client, express_cache) = default_express_cache();
+        let runtime = ThreadPool::builder().pool_size(1).create().unwrap();
+        let cache = MultilevelDataCache::new(disk_cache, express_cache, runtime);
+
+        let data = ChecksummedBytes::new("Foo".into());
+        let object_size = data.len();
+        let cache_key = ObjectId::new("a".into(), ETag::for_tests());
+
+        // put in both caches
+        cache
+            .put_block(cache_key.clone(), 0, 0, data.clone(), object_size)
+            .await
+            .expect("put should succeed");
+
+        // check it was put to express
+        assert_eq!(client.objects_number(), 1);
+
+        // check it was put to local
+        client.remove_all_objects();
+        let entry = cache
+            .get_block(&cache_key, 0, 0, object_size)
+            .await
+            .expect("cache should be accessible")
+            .expect("cache entry should be returned");
+        assert_eq!(
+            data, entry,
+            "cache entry returned should match original bytes after put"
+        );
+    }
+
+    #[tokio::test]
+    async fn test_put_from_express_to_local() {
+        let disk_cache = default_disk_cache();
+        let (client, express_cache) = default_express_cache();
+
+        let data = ChecksummedBytes::new("Foo".into());
+        let object_size = data.len();
+        let cache_key = ObjectId::new("a".into(), ETag::for_tests());
+        express_cache
+            .put_block(cache_key.clone(), 0, 0, data.clone(), object_size)
+            .await
+            .expect("put should succeed");
+
+        let runtime = ThreadPool::builder().pool_size(1).create().unwrap();
+        let cache = MultilevelDataCache::new(disk_cache, express_cache, runtime);
+
+        // get from express, put entry in the local cache
+        let entry = cache
+            .get_block(&cache_key, 0, 0, object_size)
+            .await
+            .expect("cache should be accessible")
+            .expect("cache entry should be returned");
+        assert_eq!(
+            data, entry,
+            "cache entry returned should match original bytes after put"
+        );
+
+        // delete entry from express
+        client.remove_all_objects();
+
+        // get entry from the local cache (with retries as it is async)
```

Comment:
We control the runtime, though. Could we do better?

Reply:
We can keep a reference to [ThreadPool](https://docs.rs/futures/latest/futures/executor/struct.ThreadPool.html), but I am not sure how may it help. We can not drop the thread pool or the multilevel cache, as later we need to make a `get_block` call which requires both. 

As @muddyfish suggested [below](https://github.com/awslabs/mountpoint-s3/pull/1064/files/8d9efd66ec6264356979f290b8450c1d8860168c#r1808709156) a way to make it more robust here would be to return a handle to the task from the `get_block`, but I am not sure if it is a better way compared to retries.

Reply:
No, we would need a different runtime where we could control/block on task execution, maybe built on top of a [LocalPool](https://docs.rs/futures/latest/futures/executor/struct.LocalPool.html). May not be worth just to avoid the loop here, though.